hydromt_delft3dfm/dflowfm.py

```diff
@@ -3790,70 +3571,110 @@ def get_model_time(self):
+        overwrite_grid: bool, optional
+            If True, overwrite the grid with the same name as the grid in self.mesh.
+        """
+        # Check if new grid_name
+        if grid_name not in self.mesh_names:
+            new_grid = True
+        else:
+            new_grid = False
+
+        # First call super method to add mesh data
+        super().set_mesh(
+            data=data,
+            name=name,
+            grid_name=grid_name,
+            overwrite_grid=overwrite_grid,
+        )
+
+        # check if 1D and 2D and and 1D2D links and overwrite then send warning that setup_link1d2d should be run again
+        if overwrite_grid and "link1d2d" in self.mesh.data_vars:
+            if grid_name == "mesh1d" or grid_name == "mesh2d":
+                # TODO check if warning is enough or if we should remove to be sure?
+                self.logger.warning(
+                    f"{grid_name} grid was updated in self.mesh. "
+                    "Re-run setup_link1d2d method to update the model 1D2D links."
+                )
+
+        # update related geoms if necessary: region - boundaries
+        # the region is done in hydromt core
+        if overwrite_grid or new_grid:
+            # 1D boundaries
+            if grid_name == "mesh1d":
+                self.set_geoms(self.get_boundaries(), "boundaries")
+
+    def set_link1d2d(
+        self,
+        link1d2d: xr.Dataset(),
+    ):
+        """
+        Add or replace the link1d2d in the model mesh.
+
+        Parameters
+        ----------
+        link1d2d: xr.Dataset
+            link1d2d dataset with variables: [link1d2d, link1d2d_ids, link1d2d_long_names, link1d2d_contact_type]
+        """
+        # Check if link1d2d already in self.mesh
+        if "link1d2d" in self.mesh.data_vars:
+            self.logger.info("Overwriting existing link1d2d in self.mesh.")
+            self._mesh = self._mesh.drop_vars(
+                [
+                    "link1d2d",
+                    "link1d2d_ids",
+                    "link1d2d_long_names",
+                    "link1d2d_contact_type",
+                ]
+            )
+            # Remove unused dims nLink1D2D_edge and Two
+            self._mesh = self._mesh.drop_dims(["nLink1D2D_edge", "Two"])
+
+        # Add link1d2d to mesh
+        self._mesh = self._mesh.merge(link1d2d)
 
-    @property
     def _model_has_2d(self):
-        if not self.mesh2d.is_empty():
+        if "mesh2d" in self.mesh_names:
             return True
         else:
             return False
 
-    @property
     def _model_has_1d(self):
-        if not self.mesh1d.is_empty():
+        if "mesh1d" in self.mesh_names:
             return True
         else:
             return False
+
+    def _check_crs(self):
+        """"""
+        if self.crs is None:
+            raise ValueError(
+                "CRS is not defined. Please define the CRS in the [global] init attributes before setting up the mesh."
+            )
```

Comment:
Because a default 4326 is given in the __ini__, _check_crs is not needed or is better done in the initialization.

Reply:
Removed 4326 and set default to None. _check_crs in initialisation in writing mode, and in read (from mesh) in reading mode